apps/consulting/posts/improvements-to-bsr-matrix-multiplication-in-pytorch.md

```diff
@@ -0,0 +1,292 @@
+not the product of `torch.compile`.'
+category: [Open Source Software, PyTorch, Triton, GPU]
+---
+<base target="_blank" />
+
+Quansight engineers have implemented several new kernels for matrix multiplication involving sparse
+and dense arguments. These implementations, available in PyTorch 2.1,  offer a significant
+performance improvement over those available in previous versions of PyTorch. In addition, they have
+the unique property that they are implemented using `triton`, however they are not the product of
+`torch.compile`.
+
+In this post, we will discuss the results of this work and how it can be used.
+
+## Sparse Layouts in PyTorch
+
+PyTorch implements a total of 5 sparse layouts. Each layout has properties which make it more, or
+less, suitable for a particular task. As an example, the coordinate format ([COO][pytorch-docs-coo])
+can be used to incrementally build sparse tensors, with features allowing for individual element
+access and updates to be more efficient.  However for most mathematical operations 
+[compressed layouts][pytorch-docs-compressed], like compressed sparse row 
+([CSR][pytorch-docs-csr]), are a better choice since they store specified elements with a more
+regular structure. This is what a PyTorch tensor looks like using the CSR layout
+
+```python
+>>> mat = torch.tensor([
+    [0, 0, 1, 0, 0, 0],
+    [1, 0, 1, 0, 1, 0],
+    [0, 0, 0, 1, 0, 0],
+    [0, 0, 0, 0, 0, 0],
+    [1, 1, 0, 0, 1, 0],
+    [0, 0, 0, 0, 0, 0],
+])
+>>> mat.to_sparse_csr()
+tensor(crow_indices=tensor([0, 1, 4, 5, 5, 8, 8]),
+       col_indices=tensor([2, 0, 2, 4, 3, 0, 1, 4]),
+       values=tensor([1, 1, 1, 1, 1, 1, 1, 1]),
+       size=(6, 6),
+       nnz=8,
+       layout=torch.sparse_csr)
+```
+
+This work is concerned with the block sparse row ([BSR][pytorch-docs-bsr]) layout, which is closely
+related to the CSR layout except that blocks of values are stored. This affords many of the same
+advantages as the CSR layout, with the additional feature that vectorized operations can be used to
+operate on these value blocks. Reusing the example above with the BSR layout the similarities are
+clear
+
+```python
+>>> mat.to_sparse_bsr((2, 2))
+tensor(crow_indices=tensor([0, 3, 4, 6]),
+       col_indices=tensor([0, 1, 2, 1, 0, 2]),
+       values=tensor([[[0, 0],
+                       [1, 0]],
+                        ...
+                      [[1, 0],
+                       [0, 0]]]),
+      size=(6, 6), nnz=6, layout=torch.sparse_bsr)
+```
+
+There are some drawbacks to a block sparse layout, namely we must fully materialize a block if any
+value within it is non-zero. Larger blocks are generally better for vectorized operations, but will
+have more memory overhead.  The example we have used here demonstrates this well. In
+[Figure-1][#fig-1] we have this matrix with non-zero values in yellow, and the 2x2 block size shown
+with the red grid. 
```

Comment:
This will need reflow.

```suggestion
with the red grid. The large number of red blocks that contain mostly zero values illustrates this relatively poor memory efficiency of the layout.
```